storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -241,64 +255,54 @@ static DynamicState prepareForNewAssignmentNoWorkersRunning(DynamicState dynamic
                            .withState(MachineState.WAITING_FOR_BLOB_LOCALIZATION);
     }
 
-    /**
-     * Kill the current container and start downloading what the new assignment needs, if there is a new assignment.
-     * PRECONDITION: container != null
-     * @param dynamicState current state
-     * @param staticState static data
-     * @return the next state
-     * @throws Exception
-     */
-    static DynamicState killContainerForChangedAssignment(DynamicState dynamicState, StaticState staticState) throws Exception {
-        assert (dynamicState.container != null);
+    private static DynamicState killContainerFor(KillReason reason, DynamicState dynamicState, StaticState staticState)
+            throws Exception {
+        assert dynamicState.container != null;
 
-        staticState.iSupervisor.killedWorker(staticState.port);
-        dynamicState.container.kill();
-        Future<Void> pendingDownload = null;
-        if (dynamicState.newAssignment != null) {
-            pendingDownload = staticState.localizer.requestDownloadTopologyBlobs(dynamicState.newAssignment, staticState.port,
-                                                                                 staticState.changingCallback);
+        //Skip special case if `storm kill_workers` is already invoked
+        Boolean isDead = dynamicState.container.areAllProcessesDead();
+        if (!isDead) {
+            //TODO: Is this necessary?
```

Comment:
We probably want to inform iSupervisor of the kill each time.

Reply:
I noticed that comments in the iSupervisor said the purpose of `killedWorker()` is actually to send a "task finished" update. If this is the case then we shouldn't inform iSupervisor when task halts abnormally.

Reply:
OK

Reply:
Okay. I just want to make sure that this is indeed the desired behavior.